Source/WebCore/css/parser/CSSPropertyParser.cpp

```diff
@@ -763,6 +763,105 @@ bool CSSPropertyParser::consumeFontVariantShorthand(bool important)
     return true;
 }
 
+bool CSSPropertyParser::consumeTextSpacing(bool important)
+{
+    // text-spacing: auto | none | <text-space-autospace> || <text-spacing-trim>
+    // text-autospace: auto | no-autospace
+    // text-spacing-trim: auto | space-all
+    // FIXME: parse remaining values once spec is stable.
+
+    // Part I : auto | none
+    auto firstValue = consumeIdent<CSSValueNone, CSSValueAuto>(m_range);
+    if (firstValue && m_range.atEnd()) {
+        if (m_range.atEnd()) {
+            if (firstValue->valueID() == CSSValueAuto) {
+                addProperty(CSSPropertyTextAutospace, CSSPropertyTextSpacing, firstValue.releaseNonNull(), important);
+                addProperty(CSSPropertyTextSpacingTrim, CSSPropertyTextSpacing, CSSPrimitiveValue::create(CSSValueAuto), important);
+            } else if (firstValue->valueID() == CSSValueNone) {
+                addProperty(CSSPropertyTextAutospace, CSSPropertyTextSpacing, CSSPrimitiveValue::create(CSSValueNoAutospace), important);
+                addProperty(CSSPropertyTextSpacingTrim, CSSPropertyTextSpacing, CSSPrimitiveValue::create(CSSValueSpaceAll), important);
+            }
+            // FIXME: add support for CSSValueNormal as firstValue here.
+            return true;
+        }
+    }
+    // part II (longhands): <text-space-autospace> || <text-spacing-trim> (with fist token = auto or none)
+    if (firstValue && !m_range.atEnd()) {
```

Comment:
This seems wrong, the grammar:
`normal | none | auto | <'text-autospace'> || <'text-spacing-trim'>`
should read as:

`normal | none | auto | [ <'text-autospace'> || <'text-spacing-trim'> ]`

according to https://drafts.csswg.org/css-values-4/#component-combinators

Reply:
we are parsing `normal | none | auto | [ <'text-autospace'> || <'text-spacing-trim'> ]` indeed, being discussed on pm